src/test_helpers/xds.rs

```diff
@@ -0,0 +1,179 @@
+use hyper::service::make_service_fn;
+use log::info;
+use tokio::sync::{watch, mpsc};
+use tokio_stream::wrappers::ReceiverStream;
+use tonic::client::GrpcService;
+use tonic::{Streaming, Status, Response};
+use tracing::warn;
+
+use crate::admin;
+use crate::metrics::Metrics;
+use crate::xds::{AdsClient, self};
+use crate::xds::service::discovery::v3::aggregated_discovery_service_server::{AggregatedDiscoveryServiceServer, AggregatedDiscoveryService};
+use crate::xds::service::discovery::v3::{DiscoveryResponse, DeltaDiscoveryResponse, DiscoveryRequest, DeltaDiscoveryRequest};
+use crate::{
+    identity::Identity,
+    tls,
+};
+
+/// AdsServer provides a fake ADS server implementation. Mocked responses can be assigned to it.
+#[derive(Clone)]
+pub struct AdsServer {
+    response: watch::Receiver<Result<DiscoveryResponse, tonic::Status>>,
+    delta_response: watch::Receiver<Result<DeltaDiscoveryResponse, tonic::Status>>,
+}
+impl AdsServer {
+    pub async fn spawn(cfg: xds::Config, metrics: Arc<Metrics>, block_ready: admin::BlockReady) -> (
+        watch::Sender<Result<DiscoveryResponse, tonic::Status>>,
+        watch::Sender<Result<DeltaDiscoveryResponse, tonic::Status>>,
+        AdsClient,
+    ) {
+        let delta_default = Err(tonic::Status::not_found("mock not set"));
+        let (delta_tx, delta_rx) = watch::channel(delta_default);
+
+        let default = Err(tonic::Status::not_found("mock not set"));
+        let (tx, rx) = watch::channel(default);
+
+        let server = AdsServer { response: rx, delta_response: delta_rx };
+        let listener = tokio::net::TcpListener::bind("127.0.0.1:15012").await.unwrap();
+        let certs = tls::generate_test_certs(
+            &Identity::default(),
+            Duration::from_secs(0),
+            Duration::from_secs(100),
+        );
+        let acceptor = tls::ControlPlaneCertProvider(certs);
+        let tls_stream = crate::hyper_util::tls_server(acceptor, listener);
+        let incoming = hyper::server::accept::from_stream(tls_stream);
+
+        let srv = AggregatedDiscoveryServiceServer::new(server);
+        tokio::spawn(async move {
+            hyper::Server::builder(incoming)
+                .serve(make_service_fn(move |_| {
+                    let mut srv = srv.clone();
+                    future::ok::<_, Infallible>(tower::service_fn(
+                        move |req: hyper::Request<hyper::Body>| srv.call(req),
+                    ))
+                }))
+                .await
+                .unwrap()
+        });
+        let xds_client = cfg.build(metrics, block_ready);
+        (tx, delta_tx, xds_client)
+    }
+}
+
+
+#[async_trait]
+impl AggregatedDiscoveryService for AdsServer {
+    type DeltaAggregatedResourcesStream = Pin<Box<dyn Stream<Item = Result<DeltaDiscoveryResponse, Status>> + Send>>;
+    type StreamAggregatedResourcesStream = Pin<Box<dyn Stream<Item = Result<DiscoveryResponse, Status>> + Send>>;
+    async fn stream_aggregated_resources(
```

Comment:
we don't use this in ztunnel, so we can just return an error or panic or something